usr/src/lib/libzfs/common/libzfs_dataset.c

```diff
@@ -854,16 +858,18 @@ libzfs_mnttab_find(libzfs_handle_t *hdl, const char *fsname,
 			return (ENOENT);
 	}
 
+	(void) mutex_lock(&hdl->libzfs_mnttab_cache_lock);
 	if (avl_numnodes(&hdl->libzfs_mnttab_cache) == 0)
 		libzfs_mnttab_update(hdl);
 
 	find.mtn_mt.mnt_special = (char *)fsname;
 	mtn = avl_find(&hdl->libzfs_mnttab_cache, &find, NULL);
 	if (mtn) {
 		*entry = mtn->mtn_mt;
-		return (0);
+		ret = 0;
 	}
-	return (ENOENT);
+	(void) mutex_unlock(&hdl->libzfs_mnttab_cache_lock);
```

Comment:
Use mutex_exit(), the version in user land. It will do all the error checking and abort appropriately. Otherwise, you need to check the return value and blow up if it's non-zero so we abort on deadlock or unlocking a lock that we didn't own.